Bluetooth: Add a validation for le_conn_update_complete() and bt_hci_le_subrate_change_event()

swkim101 opened the pull request:
Like 2ca179c96e3e, check and warn if incoming HCI event parameters exceed the specification to help debugging.  
Although this does not strictly comply with the specification, I don't see any misuse (like div-by-zero).  
So, adding warnings seems enough at this point.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1904,6 +1904,16 @@ static void le_conn_update_complete(struct net_buf *buf)
 	LOG_DBG("status 0x%02x %s, handle %u",
 		evt->status, bt_hci_err_to_str(evt->status), handle);
 
+	if (!IN_RANGE(sys_le16_to_cpu(evt->interval), BT_HCI_LE_INTERVAL_MIN, BT_HCI_LE_INTERVAL_MAX)) {
```
I don't think we should test these values if `evt->status != 0`.

We could consider moving these checks to after we have assigned them to the conn (since this change won't prevent that anyways), so that we don't need to do the `sys_le16_to_cpu` parts twice as well

swkim101 replied:
Done. Thank you.

rugeGerritsen commented:
Should this code be behind some Kconfig symbol to avoid increasing binary size? A controller should never give values that are out of spec

Thalley commented:
> Should this code be behind some Kconfig symbol to avoid increasing binary size? A controller should never give values that are out of spec

It's a good question. 

We have many (hopefully most) where we have some checks against what we receive from the controller. 
Question is what the host should do if we receive invalid data from a controller, if anything. 

Should we trust everything from the controller?
Should we distrust everything from the controller?

It's arguably weird that we only check a subset of data from the controller

swkim101 committed:
Bluetooth: HCI: Add validation for data from controllers

Like 2ca179c, check and warn if incoming HCI event parameters exceed
the specification. This helps debugging controllers by detecting the
out-of-spec value that shouldn't appear.

Signed-off-by: Sungwoo Kim <iam@sung-woo.kim>

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2626,6 +2640,25 @@ void bt_hci_le_subrate_change_event(struct net_buf *buf)
+			LOG_WRN("peripheral_latency exceeds the valid range 0x%04x",
+				conn->le.latency);
+		}
+		if (BT_HCI_LE_CONTINUATION_NUM_MAX < conn->le.subrate.continuation_number) {
```
Why is this not
```suggestion
		if (conn->le.subrate.continuation_number > BT_HCI_LE_CONTINUATION_NUM_MAX) {
```

swkim101 replied:
Fixed, thanks.

rugeGerritsen commented:
> > Should this code be behind some Kconfig symbol to avoid increasing binary size? A controller should never give values that are out of spec
> 
> It's a good question.
> 
> We have many (hopefully most) where we have some checks against what we receive from the controller. Question is what the host should do if we receive invalid data from a controller, if anything.
> 
> Should we trust everything from the controller? Should we distrust everything from the controller?
> 
> It's arguably weird that we only check a subset of data from the controller

I agree, we should be consistent in what we trust. I would argue that controller provided values should be trusted as this is a component that requires qualification. @swkim101 , did you see cases where the controller was providing invalid values?

Thalley commented:
> I agree, we should be consistent in what we trust. I would argue that controller provided values should be trusted as this is a component that requires qualification.

Indeed. 

Arguably all checks from the controller should be behind a Kconfig option for debugging purposes (or by ASSERTS). If a controller provides us with invalid data, how do we know if any other values (even if within range) can be trusted?

Thalley reviewed (dismissed):
The checks look good to me. The decision whether to guard the checks with a Kconfig is outside the scope of this PR

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1929,6 +1929,20 @@ static void le_conn_update_complete(struct net_buf *buf)
 			conn->le.latency = sys_le16_to_cpu(evt->latency);
 			conn->le.timeout = sys_le16_to_cpu(evt->supv_timeout);
 
+			if (!IN_RANGE(conn->le.interval, BT_HCI_LE_INTERVAL_MIN,
+				      BT_HCI_LE_INTERVAL_MAX)) {
+				LOG_WRN("interval exceeds the valid range 0x%04x",
+					conn->le.interval);
+			}
+			if (conn->le.latency > BT_HCI_LE_PERIPHERAL_LATENCY_MAX) {
+				LOG_WRN("latency exceeds the valid range 0x%04x", conn->le.latency);
+			}
+			if (!IN_RANGE(conn->le.timeout, BT_HCI_LE_SUPERVISON_TIMEOUT_MIN,
+				      BT_HCI_LE_SUPERVISON_TIMEOUT_MAX)) {
+				LOG_WRN("supv_timeout exceeds the valid range 0x%04x",
+					conn->le.timeout);
+			}
+
```
Missing handling of `CONFIG_BT_CONN_PARAM_ANY`

Thalley replied:
Good point. These checks should only be applied if `CONFIG_BT_CONN_PARAM_ANY=n`

swkim101 commented:
> > > Should this code be behind some Kconfig symbol to avoid increasing binary size? A controller should never give values that are out of spec

I agree. It looks `BT_CONN_PARAM_ANY` is designed for this. I'll add this.

> > 
> > 
> > It's a good question.
> > We have many (hopefully most) where we have some checks against what we receive from the controller. Question is what the host should do if we receive invalid data from a controller, if anything.
> > Should we trust everything from the controller? Should we distrust everything from the controller?
> > It's arguably weird that we only check a subset of data from the controller
> 
> I agree, we should be consistent in what we trust. I would argue that controller provided values should be trusted as this is a component that requires qualification. @swkim101 , did you see cases where the controller was providing invalid values?

I'm aggressively testing the BT host stack. My testing controller, which is not qualified, provided invalid values.  
However, Zephyr has accepted the out-of-spec value with this Kconfig.

https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/subsys/bluetooth/host/Kconfig#L305-L310

And it is used here:

https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/subsys/bluetooth/host/hci_core.c#L1799-L1804

Now, I need some clarification.
Let's say qualified controllers do not send out-of-spec values, and we trust controllers, i.e., assume all controllers are qualified. Then, we don't need the `BT_CONN_PARAM_ANY` option.
However, `BT_CONN_PARAM_ANY` is there, implying that (1) qualification allowed sending out-of-spec values, or (2) we do not trust controllers under `BT_CONN_PARAM_ANY`.

I think (2) more makes sense to me.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1929,6 +1929,22 @@ static void le_conn_update_complete(struct net_buf *buf)
 			conn->le.latency = sys_le16_to_cpu(evt->latency);
 			conn->le.timeout = sys_le16_to_cpu(evt->supv_timeout);
 
+#if !defined(CONFIG_BT_CONN_PARAM_ANY)
```
```suggestion
if (!IS_ENABLED(CONFIG_BT_CONN_PARAM_ANY)) {
...
}
```

cvinayak commented:
> (2) we do not trust controllers under BT_CONN_PARAM_ANY.

Yes (agree), can not trust as the values and their units used by vendors are unknown (and different range too).
